flink-connectors/flink-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSinkFunctionBase.java

```diff
@@ -0,0 +1,332 @@
+		}
+		isRunning = true;
+	}
+
+	@Override
+	public void invoke(IN value, Context context) throws Exception {
+		Mutation mutation = extract(value);
+		long mutationSize = mutation.heapSize();
+		if (batchFlushEnable) {
+			if (estimateSize != 0 && (estimateSize + mutationSize > batchFlushMaxSizeInBits || mutaionBuffer.size() + 1 > batchFlushMaxMutations)) {
+				synchronized (lock){
+					if (estimateSize != 0 && (estimateSize + mutationSize > batchFlushMaxSizeInBits
+						|| mutaionBuffer.size() + 1 > batchFlushMaxMutations)) {
+						long start = System.currentTimeMillis();
+						Exception testException = null;
+						try {
+							flushToHBaseWithRetryAndTimeout();
+						} catch (Exception e) {
+							testException = e;
+						}
+						long end = System.currentTimeMillis();
+						log.debug("Flush tasks " + (end - start) + " milliseconds with exception: " + testException);
+					}
+				}
+			}
+			synchronized (lock) {
+				mutaionBuffer.add(mutation);
+				estimateSize += mutation.heapSize();
+			}
+		} else if (mutation instanceof Put){
+			hTable.put((Put) mutation);
+		} else if (mutation instanceof Delete) {
+			hTable.delete((Delete) mutation);
+		} else if (mutation instanceof Append) {
+			hTable.append((Append) mutation);
+		} else if (mutation instanceof Increment) {
+			hTable.increment((Increment) mutation);
+		}
+	}
+
+	protected abstract Mutation extract(IN value);
+
+	protected Put generatePutMutation(IN value) {
+		byte[] rowKey = HBaseUtils.serialize(fieldTypes[rowKeyIndex], produceElementWithIndex(value, rowKeyIndex));
+		Put put = new Put(rowKey);
+		for (int i = 0; i < fieldNames.length; i++) {
+			if (i != rowKeyIndex) {
+				Object fieldValue = produceElementWithIndex(value, i);
+				if (fieldValue != null) {
+					put.addColumn(columnFamilies[i].getBytes(), qualifiers[i].getBytes(), HBaseUtils.serialize(fieldTypes[i], fieldValue));
+				}
+			}
+		}
+		return put;
+	}
+
+	protected Delete generateDeleteMutation(IN value) {
+		byte[] rowKey = HBaseUtils.serialize(fieldTypes[rowKeyIndex], produceElementWithIndex(value, rowKeyIndex));
+		Delete delete = new Delete(rowKey);
+		for (int i = 0; i < fieldNames.length; i++) {
+			if (i != rowKeyIndex) {
+				delete.addColumn(columnFamilies[i].getBytes(), qualifiers[i].getBytes());
+			}
+		}
+		return delete;
+	}
+
+	protected abstract Object produceElementWithIndex(IN value, int index);
+
+	// The HBase client operation timeout doesn't include the time of getting server state from zookeeper.
+	private void flushToHBaseWithRetryAndTimeout() throws ExecutionException, InterruptedException {
+		long start = System.currentTimeMillis();
+		try {
+			CompletableFuture<Void> flushFuture = FutureUtils
+				.retry(() -> CompletableFuture.runAsync(() -> flushToHBase(), flushExecutor), batchFlushMaxRetries, flushExecutor);
+			if (!flushFuture.isDone()) {
+				log.debug("start to wait for flush futre");
+				flushFuture.get(batchFlushMaxTimeoutMillis, TimeUnit.MILLISECONDS);
```

Comment:
It seems that we start a thread to async write data to hbase, and then block wait on the current thread. I think the async write is not needed here.